ENI: fix calculateExcessIPs excessive calculate of excess ip

wu0407 opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

<!-- Description of change -->
If usedIPs+preAllocate not over minAllocate + maxAboveWatermark and availableIPs over minAllocate + maxAboveWatermark, only care  IPs out of minAllocate + maxAboveWatermark calculateExcessIPs.

Fixes: #27970

```release-note
ENI: fix calculateExcessIPs excessive calculate of excess ip
```

gandro commented on Documentation/network/concepts/ipam/eni.rst:
```diff
@@ -308,15 +308,28 @@ is performed at the interval based scan.
 When determining whether a node has a deficit in IP addresses, the following
 calculation is performed:
 
+``availableIPs``: all IPs on ENI contain used IPs and not used IPs.
```
Why can't we not use `len(spec.ipam.available)` instead?

wu0407 replied:
The `spec.ipam` field does not contain the `available` attribute. https://github.com/wu0407/cilium/blob/8c77e5b059a80d7a1bcf9a213437ec98cc8ae4d3/pkg/ipam/types/types.go#L121

gandro replied:
Ah true, this should be `len(spec.ipam.pool)`, right?

Maybe we can do `availableIP := len(spec.ipam.pool)` to make it a bit more readable

wu0407 replied:
yeah

gandro commented on Documentation/network/concepts/ipam/eni.rst:
```diff
@@ -308,15 +308,28 @@ is performed at the interval based scan.
 .. code-block:: go
 
-     spec.ipam.pre-allocate - (len(spec.ipam.available) - len(status.ipam.used))
+     needIPs = max(spec.ipam.pre-allocate - availableIPs - len(status.ipam.used), spec.ipam.min-allocate - availableIPs)
```
nit:

```suggestion
     neededIPs = max(spec.ipam.pre-allocate - availableIPs - len(status.ipam.used), spec.ipam.min-allocate - availableIPs)
```

gandro reviewed (commented):
Looks good, one question

gandro commented on Documentation/network/concepts/ipam/eni.rst:
```diff
@@ -308,15 +308,29 @@ is performed at the interval based scan.
 
-     spec.ipam.pre-allocate - (len(spec.ipam.available) - len(status.ipam.used))
+     availableIP := len(spec.ipam.pool)
+     neededIPs = max(spec.ipam.pre-allocate - availableIPs - len(status.ipam.used), spec.ipam.min-allocate - availableIPs)
```
Shouldn't this be the following?

```suggestion
     neededIPs = max(spec.ipam.pre-allocate - (availableIPs - len(status.ipam.used)), spec.ipam.min-allocate - availableIPs)
```

gandro reviewed (commented):
Thanks! One last thing that I noticed, then this should be good!

learnitall commented on Documentation/network/concepts/ipam/eni.rst:
```diff
@@ -308,15 +308,29 @@ is performed at the interval based scan.
 When determining whether a node has a deficit in IP addresses, the following
 calculation is performed:
 
+
```
Do you mind removing this newline?

learnitall commented on Documentation/network/concepts/ipam/eni.rst:
```diff
@@ -308,15 +308,29 @@ is performed at the interval based scan.
 .. code-block:: go
 
-     spec.ipam.pre-allocate - (len(spec.ipam.available) - len(status.ipam.used))
+     availableIP := len(spec.ipam.pool)
```
```suggestion
     availableIPs := len(spec.ipam.pool)
```

learnitall commented on Documentation/network/concepts/ipam/eni.rst:
```diff
@@ -308,15 +308,29 @@ is performed at the interval based scan.
 .. code-block:: go
 
-     (len(spec.ipam.available) - len(status.ipam.used)) - (spec.ipam.pre-allocate + spec.ipam.max-above-watermark)
+     availableIP := len(spec.ipam.pool)
```
```suggestion
     availableIPs := len(spec.ipam.pool)
```

learnitall reviewed (changes requested):
There seem to be a couple of typos in the docs, but otherwise looks good!

wu0407 committed:
ENI: fix calculateExcessIPs excessive calculate of excess ip

If usedIPs+preAllocate not over minAllocate + maxAboveWatermark, only care in
all ips out of minAllocate + maxAboveWatermark calculateExcessIPs.

Fixes #27970

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>

wu0407 committed:
Update formulas that need IP and exceed IP for calculation on eni docs

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>
